ONSdigital/eq-author-app
# ✨ EAR 1315 - Add API resolver to import questions from another questionnaire

### What is the context of this PR?

[Ticket](https://collaborate2.ons.gov.uk/jira/browse/EAR-1315)

Adds `importQuestions` resolver  which imports questions specified by `questionIds` from questionnaire with ID `questionnaireId` at the provided `Position` (specified by `folderId`, `sectionId`, and the `index` to which the items should be inserted from).

Resolver mimics `movePage` position handling in that you can specify either a `folderId` or `sectionId`, `folderId` (being more specific) takes higher precedence. The `index` specifies the position to start inserting from in *either* a section (if `sectionId` and not `folderId` provided) or folder (if `folderId` provided).

`UserInputErrors` are thrown for any possible failure conditions to do with invalid input.

The input questions are mapped to imported questions like so:
- All IDs are remapped
- All nested `qCode` and `secondaryQCode` attributes are set to `null`
- Piping is *not* deleted for now after conversation with Joe - this will preserve the `piping deleted` errors when questions are imported (since IDs will not be resolvable within the questionnaire) for purpose of the initial MVP
- If the target is a `Section`, each imported question is wrapped in its own disabled folder per ticket spec.

The resolver returns `Section` for now (the section modified in the target questionnaire) for ease of integrating the response on the client, e.g. utilising automatic cache updates (idiomatic graphql would probably be to return only the folder(s) of imported content, requiring manual cache writes - can do that instead if you want).

### How to review

- Make a source questionnaire with a set of `questionIds` - make a note of these
- Make a target questionnaire
- While on the target questionnaire:
- Use Apollo dev tools to execute a query like the following:

```graphql
mutation ImportStuff($input: ImportQuestionsInput!) {
    importQuestions(input: $input) {
      id
      <...whatever else you want from Section>
   }
})
```

and as input:

```graphql
input: {
  questionIds: [<question IDs from source>],
  questionnaireId: <source questionnaire ID>,
  position: {
    sectionId: <target section ID in current questionnaire>,
    index: <index you'd like questions to be inserted into in target section>
  }
}
```
- Should get the output you'd expect from the above
- Refresh the page & you should see the new imported content
- qCodes should be empty for imported stuff
- If you want, try the same thing but with a target `folderId` instead

### What to do after everything is green

1. - [ ] Bring this branch up-to-date with Origin/Master
2. - [ ] If there are a lot of commits or some are not helpful to read, squash them down
3. - [ ] Click the **Merge** button on this pull request
4. - [ ] Does this change mean the **Capability examples** questionnaire should be updated?
5. - [ ] Move the Jira ticket for this task into the next stage of the process
